### PR TITLE
Reduce duplicate code in receipts servlets.

### PR DESCRIPTION
**Pull request by clokep**
I was poking at this while looking at receipts and I found this be a nice clean-up, so figured I should put it in a separate PR.

Factors both the read marker and receipts servlets a bit:

* Pre-calculates the allowed receipt types (instead of on each request).
* For the read marker servlet, iterate over the items in the body instead of again calculating which receipts are valid.

### CODE REVIEW

File: synapse/rest/client/read_marker.py [hunk 1]
```diff
@@ -65,31 +63,23 @@ async def on_POST(
...
-        read_marker_event_id = body.get(ReceiptTypes.FULLY_READ, None)
-        if read_marker_event_id:
-            await self.read_marker_handler.received_client_read_marker(
-                room_id,
-                user_id=requester.user.to_string(),
-                event_id=read_marker_event_id,
-            )
+        for receipt_type, event_id in body.items():
+            if not event_id:
+                continue
```
Comment (clokep) [inline synapse/rest/client/read_marker.py:68]: This is here to keep the logic the same as previous -- if you had a weird values in the body which became false-y than Synapse would ignore them.

We can increase validation here but that's a bit orthogonal to what this PR is trying to do.

File: synapse/rest/client/receipts.py [hunk 2]
```diff
@@ -39,31 +39,27 @@ class ReceiptRestServlet(RestServlet):
...
-        if self.hs.config.experimental.msc2285_enabled and receipt_type not in [
-            ReceiptTypes.READ,
-            ReceiptTypes.READ_PRIVATE,
-            ReceiptTypes.FULLY_READ,
-        ]:
+        if receipt_type not in self._known_receipt_types:
             raise SynapseError(
                 400,
-                "Receipt type must be 'm.read', 'org.matrix.msc2285.read.private' or 'm.fully_read'",
+                f"Receipt type must be {', '.join(self._known_receipt_types)}",
```
Comment (clokep) [inline synapse/rest/client/receipts.py:61]: This error message isn't quite as good, but I think it is fine still.

**Comment by clokep** [comment]
`TestSendJoinPartialStateResponse` is failing, which is a known flake (https://github.com/matrix-org/synapse/issues/13161#issuecomment-1175040236)

File: synapse/rest/client/read_marker.py [hunk 3]
```diff
@@ -65,31 +63,23 @@ async def on_POST(
...
-            )
-
-        read_marker_event_id = body.get(ReceiptTypes.FULLY_READ, None)
-        if read_marker_event_id:
-            await self.read_marker_handler.received_client_read_marker(
-                room_id,
-                user_id=requester.user.to_string(),
-                event_id=read_marker_event_id,
-            )
+        for receipt_type, event_id in body.items():
```
Comment (H-Shay) [inline synapse/rest/client/read_marker.py:66]: Will this loop process unrecognized receipt types if they are in the body?

**Review by H-Shay — APPROVED** [review]

File: synapse/rest/client/read_marker.py (hunk 3 above)
Comment (clokep) [reply synapse/rest/client/read_marker.py:66]: No, a few lines above this we assert that all the keys (the receipt types) are known.

Comment (clokep) [reply synapse/rest/client/read_marker.py:66]: Oh, hmm.... looks like we don't do that as I thought we did, heh. That just logs, doesn't error. Good catch!

**Comment by clokep** [comment]
@H-Shay Can you take another look at this? Thanks!

**Comment by H-Shay** [comment]
I just noticed that there's more than one complement failure and looks like they are related to sync-are these known flakes?

**Comment by clokep** [comment]
I'll re-run it, I don't see it listed in #13161

It was `TestSync/parallel/sync_should_succeed_even_if_the_sync_token_points_to_a_redaction_of_an_unknown_event` for the record.

**Comment by H-Shay** [comment]
I just saw this same complement test fail on an unrelated PR of mine, I think it's a flake. I will add it to the list.

**Comment by clokep** [comment]
> I just saw this same complement test fail on an unrelated PR of mine, I think it's a flake. I will add it to the list.

Thanks! Re-running them again did pass, so should be ready for another look!

**Review by H-Shay — APPROVED** [review]

**Pull request merged**